### ISSUE TITLE
[BUG] Version 0.5.0 uses 99% of CPU on Archer C7v2

### ISSUE THREAD

**Issue by richb-hanover** [sqm-autorate/sqm-autorate#139]
**Script Version** 0.5.0

**Describe the bug**
I had trouble getting `sqm-autorate` to go more than about half my ISP's speed when I installed it on an Archer C7v2. I did a bunch of debugging and found that `sqm-autorate.lua` was sometimes using 99% of the CPU:

```
Mem: 56772K used, 66628K free, 1252K shrd, 3344K buff, 18244K cached
CPU:  93% usr   1% sys   0% nic   0% idle   0% io   0% irq   4% sirq
Load average: 1.87 1.56 1.51 3/57 6469
  PID  PPID USER     STAT   VSZ %VSZ %CPU COMMAND
 6307     1 root     S    10240   8%  99% lua sqm-autorate.lua
...
```

**To Reproduce**
1. Install 0.5.0 from the `testing/next` branch
2. Let it run
3. I cannot download or upload more than ~12-13mbps
4. Speed keeps getting knocked back to the 50% value

**OpenWrt Details (please paste your output from `cat /etc/os-release`):**
 - OS-Release: 21.02.1 on Archer C7v2
 - Connection Type + Speed: 25/25 symmetric

Debugging info and files:
[openwrtstats-2022-02-02-215254.gz](https://github.com/sqm-autorate/sqm-autorate/files/7991493/openwrtstats-2022-02-02-215254.gz)

**Comment by chas-iot** [comment]
Is this reproducible?

**Comment by richb-hanover** [comment]
It seems to be. I left it running overnight in a spare router, and it's still consuming 90+% of the CPU in `top' when the network is idle.

```
Mem: 71880K used, 51520K free, 12700K shrd, 3344K buff, 29692K cached
CPU:  96% usr   1% sys   0% nic   0% idle   0% io   0% irq   1% sirq
Load average: 1.61 1.63 1.60 2/57 13669
  PID  PPID USER     STAT   VSZ %VSZ %CPU COMMAND
 6307     1 root     S    13108  11%  99% lua sqm-autorate.lua
 1582     1 root     S     4288   3%   1% /usr/sbin/hostapd -s -g /var/run/hostapd/global
13595 13587 root     R     1252   1%   0% top
...
```

In my main router (TP-Link DIR-878) `sqm-autorate.lua` is using a lot less:

```
Mem: 91348K used, 30880K free, 11888K shrd, 4668K buff, 30188K cached
CPU:  23% usr   2% sys   0% nic  52% idle   0% io   0% irq  21% sirq
Load average: 1.00 0.89 0.85 3/92 10485
  PID  PPID USER     STAT   VSZ %VSZ %CPU COMMAND
 7732     1 root     S    11100   9%  26% lua sqm-autorate.lua
10485 10474 root     R     1260   1%   4% top
...
```

**Comment by dim-geo** [comment]
It happens also in turris.

**Comment by chas-iot** [comment]
I saw one other report of this 100% CPU use. @richb-hanover if you stop and restart v0.5 on Archer C7v2, does this behaviour re-occur? This is not to rule it out as a one-off but to get some idea of the shape of this issue.

**Comment by Lochnair** [comment]
I'd also be curious to know which thread(s) is responsible for the high usage. Since the set_thread_name change is in this release.

Add `COMM` to active columns in htop to see the thread names:
![image](https://user-images.githubusercontent.com/954926/153015142-1b1db535-b54b-40a3-a9a5-4c55f240a385.png)

Result:
![image](https://user-images.githubusercontent.com/954926/153015394-97e4202b-7ef6-44ee-87a2-1c1b489639a6.png)

**Comment by TaxFairness** [comment]
Sorry for taking long to get back to this...
* Yes, stopping and restarting the `sqm-autorate` service exhibits the behavior
* Here's the htop showing the thread names running on a TP-Link Archer C7 v2:

<img width="911" alt="image" src="https://user-images.githubusercontent.com/99080719/153534789-2836e0b9-8927-4c98-9d52-ead9e2564fd9.png">

* And here's regular old top...
<img width="912" alt="image" src="https://user-images.githubusercontent.com/99080719/153534927-c1df5da2-1604-46e1-8657-08033e93ee71.png">

**Comment by richb-hanover** [comment]
Thanks for this - I'm installing now. A request: When you branch from, say 0.5.0, would you update the version number, even if it's "0.5.0-develop/cpu-profiling" so that we can tell which version's running? Thanks.

**Comment by richb-hanover** [comment]
Question: As I was speculating about why the `baseliner` thread used so much of the processor, I decided to read through the code.

I note that the other threads (ping_sender, ping_receiver, peer_selector, etc) have a call to nsleep() that controls the rate they run.

It doesn't look as if `baseliner` has any calls to a sleep() - could that result in it being an "infinite loop"? (I'm entirely willing to be proven wrong here... :-)

**Comment by Lochnair** [comment]
> Thanks for this - I'm installing now. A request: When you branch from, say 0.5.0, would you update the version number, even if it's "0.5.0-develop/cpu-profiling" so that we can tell which version's running? Thanks.

I was thinking about having the setup script do that for you, if possible including the branch _and_ commit if possible. Would be rather helpful if we're not certain what exact version we're running.

> **Update #2:** After letting it run for a while (15 minutes) and running a bunch of speed tests (`betterspeedtest.sh`) on the router, it peaked as high as ~95% again.

I do still expect peaks even with my changes. That usually calms down after the peer reselector has done it's initial 40 runs with 30 sec between each, so around 20 minutes after startup till it stabilizes.

**Comment by chas-iot** [comment]
I find that the patch reduced CPU usage noticeably. Difficult to tell exactly but in the 25% to 33% range on a high-end router

**Comment by richb-hanover** [comment]
Maybe I'm muddying the waters... My reports from the last 24 hours are from the `develop/setup-config` branch.

Does that branch contain the patch that decreases CPU?

**Comment by Lochnair** [comment]
About the lack of sleep in `baseline_calculator`, it doesn't need to as it blocks here:

```lua
local _, time_data = stats_queue:receive(nil, "stats")
```

When the first parameter for the `receive` call is `nil` it waits indefinitely until there's something in the queue to get,
meaning it will wait until `ping_receiver` has processed a packet and sent the stats. So that shouldn't use much CPU time.

What I found was that the `owd_data:get("owd_tables")` call was rather expensive, and since we don't actually change those data outside of the `baseline_calculator` function, we can move it out of the loop to save CPU.

The calls to look up the actual data structures in the table also seemed expensive, so that was also moved outside the loop:
```lua
local owd_baseline = owd_tables["baseline"]
```

**Comment by richb-hanover** [comment]
> ```lua
> local _, time_data = stats_queue:receive(nil, "stats")
> ```
>
> When the first parameter for the receive call is nil it waits indefinitely until there's something in the queue to get...

OK. Newbie lua code-reader mistake. (Whew!)

I will switch back to the `develop/cpu-profiling` branch, especially on my Archer C7v2 that opened the initial report.

**Comment by richb-hanover** [comment]
Update: Using ArcherC7v2 and the `develop/cpu-profiling` branch... CPU is still high during the training period. I will report back later on (after an hour?)

<img width="828" alt="image" src="https://user-images.githubusercontent.com/1094930/154066807-16786f1f-e61c-4225-b61f-a2b51185921f.png">


NB: For my own records: I installed this branch with these commands:

```
cd /tmp
# Remove old sqm-autorate
curl https://raw.githubusercontent.com/sqm-autorate/sqm-autorate/develop/setup-config/sqm-autorate-remove.sh > remove.sh
sh remove.sh
# Verify that sqm-autorate files are not present in /tmp, /etc/config/, /etc/init.d

# Download setup.sh from develop/cpu-profiling
root@LBWO-C7v2:~# curl https://raw.githubusercontent.com/sqm-autorate/sqm-autorate/develop/cpu-profiling/sqm-autorate-setup.sh > sqm-autorate-setup.sh
sh ./sqm-autorate-setup.sh develop/cpu-profiling

# edit /etc/config/sqm-autorate to have good parameters

# start and enable the service
```

**Comment by richb-hanover** [comment]
I think I must be doing something wrong... I installed the `develop/cpu-profiling` branch (using the procedure above) and let it run. There are many problems:

- Bursts of high CPU. `top` and `htop` show that. Sometimes `htop` fails to update for many seconds, then "catches up". The LuCI web gui gets slow, and sometimes it shows "RPC Error" or some such message indicating a slow response from the router. Typing into a SSH session has bad response times (sometimes several seconds to see character echoes). 

- I cannot get more than ~15mbps from my 25mbps ISP. `tc -s qdisc` shows the CAKE settings around 15000, and the `/tmp/sqm-autorate.csv` shows those values  either not changing from my defaults, or increasing briefly and dropping back to the configured speeds.

Here are my stats...
[openwrtstats-2022-02-15-082119.gz](https://github.com/sqm-autorate/sqm-autorate/files/8069502/openwrtstats-2022-02-15-082119.gz)

Thoughts on what could be happening? Thanks.

**Comment by Lochnair** [comment]
I believe the reason for all of it is that sqm-autorate exhausts the CPU, so there's not enough to spare for anything else running. So clearly as it stands today it's not suitable to use one devices without a beefy enough CPU.

I think we can get the CPU usage by quite a bit more, but I honestly don't know if we can get it to a level I'd consider acceptable on the ArcherC7v2 or similarly specced devices

**Comment by richb-hanover** [comment]
> I believe the reason for all of it is that sqm-autorate exhausts the CPU,

Perhaps it's unavoidable, but that would be a shame. (The ArcherC7 was voted the "Best internet router" by Wirecutter not that many years ago...)

I wonder if delaying the formatting of the printf() calls when not actually logging might make a difference. That is, the code currently does all the work to format the strings for the printf() call then says, "Oops... the level is WARN, so I'll throw that result away."

Changing the logging functions to check the level first, then formatting if it's going to be displayed might make a difference. (Makes you hunger for C-style macros, doesn't it...)

**Comment by chas-iot** [comment]
closed by #148
